File: ios/Classes/Json/SuperwallEventInfo+Json.swift
import SuperwallKit

extension SuperwallEventInfo {
  func toJson() -> [String: Any] {
      return [
          "event": event.toJson(),
          "params": params
      ]
  }
}

extension SuperwallEvent {
  func toJson() -> [String: Any] {
    switch self {
      case .firstSeen:
        return ["event": "firstSeen"]
      case .appOpen:
        return ["event": "appOpen"]
      case .appLaunch:
        return ["event": "appLaunch"]
        case .identityAlias:
        return ["event": "identityAlias"]
      case .appInstall:
        return ["event": "appInstall"]
      case .sessionStart:
        return ["event": "sessionStart"]
      case .deviceAttributes(let attributes):
        return ["event": "deviceAttributes", "attributes": attributes]
      case .subscriptionStatusDidChange:
        return ["event": "subscriptionStatusDidChange"]
      case .appClose:
        return ["event": "appClose"]
      case .deepLink(let url):
        return ["event": "deepLink", "url": url.absoluteString]
      case .triggerFire(let eventName, let result):
        return ["event": "triggerFire", "eventName": eventName, "result": result.toJson()]
      case .paywallOpen(let paywallInfo):
        return ["event": "paywallOpen", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallClose(let paywallInfo):
        return ["event": "paywallClose", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallDecline(let paywallInfo):
        return ["event": "paywallDecline", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .transactionStart(let product, let paywallInfo):
        return ["event": "transactionStart", "product": product.toJson(), "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .transactionFail(let error, let paywallInfo):
        return ["event": "transactionFail", "error": error.localizedDescription, "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .transactionAbandon(let product, let paywallInfo):
        return ["event": "transactionAbandon", "product": product.toJson(), "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .transactionComplete(let transaction, let product, let paywallInfo):
        var json: [String: Any] = ["event": "transactionComplete", "product": product.toJson(), "paywallInfoBridgeId": paywallInfo.createBridgeId()]
        let transactionJson = transaction?.toJson()
        if let transactionJson {
          json["transaction"] = transactionJson
        }
        return json
      case .subscriptionStart(let product, let paywallInfo):
        return ["event": "subscriptionStart", "product": product.toJson(), "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .freeTrialStart(let product, let paywallInfo):
        return ["event": "freeTrialStart", "product": product.toJson(), "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .transactionRestore(let restoreType, let paywallInfo):
        return ["event": "transactionRestore", "restoreType": restoreType.toJson(), "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .transactionTimeout(let paywallInfo):
        return ["event": "transactionTimeout", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .userAttributes(let attributes):
        return ["event": "userAttributes", "attributes": attributes]
      case .nonRecurringProductPurchase(let product, let paywallInfo):
        return ["event": "nonRecurringProductPurchase", "product": product.toJson(), "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallResponseLoadStart(let triggeredEventName):
        return ["event": "paywallResponseLoadStart", "triggeredEventName": triggeredEventName ?? ""]
      case .paywallResponseLoadNotFound(let triggeredEventName):
        return ["event": "paywallResponseLoadNotFound", "triggeredEventName": triggeredEventName ?? ""]
      case .paywallResponseLoadFail(let triggeredEventName):
        return ["event": "paywallResponseLoadFail", "triggeredEventName": triggeredEventName ?? ""]
      case .paywallResponseLoadComplete(let triggeredEventName, let paywallInfo):
        return ["event": "paywallResponseLoadComplete", "triggeredEventName": triggeredEventName ?? "", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallWebviewLoadStart(let paywallInfo):
        return ["event": "paywallWebviewLoadStart", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallWebviewLoadFail(let paywallInfo):
        return ["event": "paywallWebviewLoadFail", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallWebviewLoadComplete(let paywallInfo):
        return ["event": "paywallWebviewLoadComplete", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallWebviewLoadTimeout(let paywallInfo):
        return ["event": "paywallWebviewLoadTimeout", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallProductsLoadStart(let triggeredEventName, let paywallInfo):
        return ["event": "paywallProductsLoadStart", "triggeredEventName": triggeredEventName ?? "", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallProductsLoadFail(let triggeredEventName, let paywallInfo):
        return ["event": "paywallProductsLoadFail", "triggeredEventName": triggeredEventName ?? "", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallProductsLoadComplete(let triggeredEventName):
        return ["event": "paywallProductsLoadComplete", "triggeredEventName": triggeredEventName ?? ""]
      case .surveyResponse(let survey, let selectedOption, let customResponse, let paywallInfo):
        return ["event": "surveyResponse", "survey": survey.toJson(), "selectedOption": selectedOption.toJson(), "customResponse": customResponse ?? "", "paywallInfoBridgeId": paywallInfo.createBridgeId()]
      case .paywallPresentationRequest(let status, let reason):
        var json: [String: Any] = ["event": "paywallPresentationRequest", "status": status.toJson()]
        let reasonJson = reason?.toJson()
        if let reasonJson {
          json["reason"] = reasonJson
        }
        return json
      case .touchesBegan:
        return ["event": "touchesBegan"]
      case .surveyClose:
        return ["event": "surveyClose"]
      case .reset:
        return ["event": "reset"]
    }
  }
}

File: android/src/main/kotlin/com/superwall/superwallkit_flutter/json/SuperwallEventInfo+Json.kt
import com.superwall.sdk.analytics.superwall.SuperwallEvent
import com.superwall.sdk.analytics.superwall.SuperwallEventInfo
import com.superwall.sdk.store.abstractions.transactions.StoreTransaction
import com.superwall.superwallkit_flutter.json.toJson

suspend fun SuperwallEventInfo.toJson(): Map<String, Any> {
    return mapOf(
        "placement" to placement.toJson(),
        "params" to params
    )
}

suspend fun SuperwallEvent.toJson(): Map<String, Any?> {
    return when (this) {
        is SuperwallEvent.FirstSeen -> mapOf("placement" to "firstSeen")
        is SuperwallEvent.AppOpen -> mapOf("placement" to "appOpen")
        is SuperwallEvent.AppLaunch -> mapOf("placement" to "appLaunch")
        is SuperwallEvent.IdentityAlias -> mapOf("placement" to "identityAlias")
        is SuperwallEvent.AppInstall -> mapOf("placement" to "appInstall")
        is SuperwallEvent.SessionStart -> mapOf("placement" to "sessionStart")
        is SuperwallEvent.Reset -> mapOf("placement" to "reset")
        is SuperwallEvent.Restore.Start -> mapOf("placement" to "restoreStart")
        is SuperwallEvent.Restore.Complete -> mapOf("placement" to "restoreComplete")
        is SuperwallEvent.Restore.Fail -> {
            mapOf("placement" to "restoreFail")
            mapOf("message" to this.reason)
        }

        is SuperwallEvent.DeviceAttributes -> mapOf(
            "placement" to "deviceAttributes",
            "attributes" to this.attributes
        )

        //is SuperwallEvent.SubscriptionStatusDidChange -> mapOf("placement" to "subscriptionStatusDidChange")
        is SuperwallEvent.AppClose -> mapOf("placement" to "appClose")
        is SuperwallEvent.DeepLink -> mapOf("placement" to "deepLink", "url" to this.uri.toString())
        is SuperwallEvent.TriggerFire -> mapOf(
            "placement" to "triggerFire",
            "placementName" to this.placementName,
            "result" to this.result.toJson()
        )

        is SuperwallEvent.PaywallOpen -> mapOf(
            "placement" to "paywallOpen",
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallClose -> mapOf(
            "placement" to "paywallClose",
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallDecline -> mapOf(
            "placement" to "paywallDecline",
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.TransactionStart -> mapOf(
            "placement" to "transactionStart",
            "product" to this.product.toJson(),
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.TransactionFail -> mapOf(
            "placement" to "transactionFail",
            "error" to this.error.localizedMessage,
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.TransactionAbandon -> mapOf(
            "placement" to "transactionAbandon",
            "product" to this.product.toJson(),
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.TransactionComplete -> {
            val json = mutableMapOf<String, Any?>(
                "placement" to "transactionComplete",
                "product" to this.product.toJson(),
                "paywallInfoBridgeId" to ""
            )
            val transaction = this.transaction as? StoreTransaction
            val transactionJson = transaction?.toJson()
            if (transactionJson != null) {
                json["transaction"] = transactionJson
            }

            json["transaction"] = transaction?.toJson()
            return json
        }

        is SuperwallEvent.SubscriptionStart -> mapOf(
            "placement" to "subscriptionStart",
            "product" to this.product.toJson(),
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.FreeTrialStart -> mapOf(
            "placement" to "freeTrialStart",
            "product" to this.product.toJson(),
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.TransactionRestore -> mapOf(
            "placement" to "transactionRestore",
            "restoreType" to this.restoreType.toJson(),
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.TransactionTimeout -> mapOf(
            "placement" to "transactionTimeout",
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.UserAttributes -> mapOf(
            "placement" to "userAttributes",
            "attributes" to this.attributes
        )

        is SuperwallEvent.NonRecurringProductPurchase -> mapOf(
            "placement" to "nonRecurringProductPurchase",
            "product" to this.product.toJson(),
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallResponseLoadStart -> mapOf(
            "placement" to "paywallResponseLoadStart",
            "triggeredPlacementName" to (this.triggeredPlacementName ?: "")
        )

        is SuperwallEvent.PaywallResponseLoadNotFound -> mapOf(
            "placement" to "paywallResponseLoadNotFound",
            "triggeredPlacementName" to (this.triggeredPlacementName ?: "")
        )

        is SuperwallEvent.PaywallResponseLoadFail -> mapOf(
            "placement" to "paywallResponseLoadFail",
            "triggeredPlacementName" to (this.triggeredPlacementName ?: "")
        )

        is SuperwallEvent.PaywallResponseLoadComplete -> mapOf(
            "placement" to "paywallResponseLoadComplete",
            "triggeredPlacementName" to (this.triggeredPlacementName ?: ""),
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallWebviewLoadStart -> mapOf(
            "placement" to "paywallWebviewLoadStart",
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallWebviewLoadFail -> mapOf(
            "placement" to "paywallWebviewLoadFail",
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallWebviewLoadComplete -> mapOf(
            "placement" to "paywallWebviewLoadComplete",
            "paywallInfoBridgeId" to ""
        )
        is SuperwallEvent.PaywallWebviewLoadFallback -> mapOf(
            "placement" to "paywallWebviewLoadFallback",
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallWebviewLoadTimeout -> mapOf(
            "placement" to "paywallWebviewLoadTimeout",
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallProductsLoadStart -> mapOf(
            "placement" to "paywallProductsLoadStart",
            "triggeredPlacementName" to (this.triggeredPlacementName ?: ""),
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallProductsLoadFail -> mapOf(
            "placement" to "paywallProductsLoadFail",
            "triggeredPlacementName" to (this.triggeredPlacementName ?: ""),
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallProductsLoadComplete -> mapOf(
            "placement" to "paywallProductsLoadComplete",
            "triggeredPlacementName" to (this.triggeredPlacementName ?: "")
        )

        is SuperwallEvent.SurveyResponse -> mapOf(
            "placement" to "surveyResponse",
            "survey" to this.survey.toJson(),
            "selectedOption" to this.selectedOption.toJson(),
            "customResponse" to (this.customResponse ?: ""),
            "paywallInfoBridgeId" to ""
        )

        is SuperwallEvent.PaywallPresentationRequest -> {
            val json = mutableMapOf(
                "placement" to "paywallPresentationRequest",
                "status" to this.status.toJson()
            )
            val reasonJson = this.reason?.toJson()
            if (reasonJson != null) {
                json["reason"] = reasonJson
            }
            return json
        }
        is SuperwallEvent.SurveyClose -> mapOf("placement" to "surveyClose")
        is SuperwallEvent.ConfigRefresh -> mapOf("placement" to "configRefresh")
        is SuperwallEvent.ConfigAttributes -> mapOf("placement" to "configAttributes")
        is SuperwallEvent.CustomPlacement -> mapOf(
            "placement" to "customPlacement",
            "name" to this.placementName,
            "params" to this.params,
            "paywallInfoBridgeId" to ""
        )
        is SuperwallEvent.ConfigFail -> mapOf("placement" to "configFail")
        is SuperwallEvent.ConfirmAllAssignments -> mapOf("placement" to "confirmAllAssignments")
        is SuperwallEvent.PaywallResourceLoadFail ->  mapOf("placement" to "paywallResourceLoadFail")
        is SuperwallEvent.ShimmerViewStart -> mapOf("placement" to "shimmerViewStart")
        is SuperwallEvent.ShimmerViewComplete ->  mapOf("placement" to "shimmerViewComplete")
        else -> {mapOf()}

    }
}